### PR TITLE
9.2/upload and rich text

### PR DESCRIPTION
**Pull request by orthagh**
Ok.

Here is the first usable and (i think) clean version for rich text fields with image inline pasting.
The feature targets tickets, followups, tasks and kb (for the moment).

My work is based on [Stonebuzz one](https://github.com/stonebuzz/glpi/tree/9.1/uploadAndRichText), and with a lot of clean on old code (from infotel).
FI, I dropped crop image feature.

I'll appreciate critical review, this feature need to be very clean.

Work incoming:
- tests with mailcollector
- some tests (pretty hard to cover here)

### CODE REVIEW

**Comment by orthagh** [comment]
Oh, my branch is from master, so conflict is imo not important.

File: ajax/fileupload.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+* @brief
+* @since version 0.85
+*/
+
+$AJAX_INCLUDE = 1;
+include ('../inc/includes.php');
+include_once(GLPI_JQUERY_UPLOADHANDLER);
+
+$errors = [
+   1                     => __('The uploaded file exceeds the upload_max_filesize directive in php.ini'),
```
Comment (trasher) [inline ajax/fileupload.php]: Should not constants used here instead of hardcoed values?

File: ajax/fileupload.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            $val->filesize = Toolbox::getSize($val->size);
+            if (isset($_REQUEST['showfilesize']) && $_REQUEST['showfilesize']) {
+               $val->display = sprintf('%1$s %2$s', $val->display, $val->filesize);
+            }
+         }
+      } else {
+         // Unlink file
+         $val->error = $errors['accept_file_types'];
+         if (file_exists($upload_dir.$val->name)) {
+            @unlink($upload_dir.$val->name);
```
Comment (trasher) [inline ajax/fileupload.php]: Should not the `@` be removed?

File: inc/knowbaseitem.class.php [hunk 3]
```diff
@@ -740,6 +749,8 @@ function showForm($ID, $options=array()) {
                                        $params);
          echo "<span id='visibility$rand'></span>";
          echo "</td></tr>\n";
+
```
Comment (trasher) [inline inc/knowbaseitem.class.php]: CS ;)

File: script.js [hunk 4]
```diff
@@ -938,3 +938,621 @@ if ($(window).width() <= 700) {
 $(function(){
    $.ui.dialog.prototype._focusTabbable = function(){};
 });
+
```
Comment (trasher) [inline script.js]: I'd vote to put the specific upload code into its own file; and load it only when required

**Review by trasher — COMMENTED** [review]
Come minor comments, I've not yet tested the changes

File: ajax/fileupload.php (hunk 1 above)
Comment (orthagh) [reply ajax/fileupload.php]: maybe, i'll think about it.
In fact, i moved (and tweaked a little) the file from front/

File: script.js (hunk 4 above)
Comment (orthagh) [reply script.js]: good point for that.
Maybe be difficult atm.
Ticket, task, followups and kb concerned.

I think it should be a more global refactor for all js library.

BTW, i could separate fileupload script in another file (but still load everywhere)

File: ajax/fileupload.php (hunk 2 above)
Comment (orthagh) [reply ajax/fileupload.php]: done

File: inc/knowbaseitem.class.php (hunk 3 above)
Comment (orthagh) [reply inc/knowbaseitem.class.php]: done

File: script.js (hunk 4 above)
Comment (wawax) [reply script.js]: Reminder should also be affected by this PR ?

Comment (trasher) [reply script.js]: OK for loading everywhere; if we do not have choice for now. But +1 to use a specific file, will be easier to load when required in the future ;)

File: ajax/fileupload.php (hunk 1 above)
Comment (trasher) [reply ajax/fileupload.php]: Yep, I've seen it's been moved; but original code should have used constants... Well; it's not a huge work, nor a blocker.

Comment (orthagh) [reply ajax/fileupload.php]: done

File: script.js (hunk 4 above)
Comment (orthagh) [reply script.js]: done

**Pull request closed without merging**